add AGENTS.md with AEM Edge Delivery project guidelines/instructions (#532)
adobe/aem-boilerplate: AGENTS.md

# AGENTS.md

> This is the AEM Edge Delivery boilerplate project - a starter template for AEM Edge Delivery Services projects. This file provides guidance for AI coding agents working on this project.

Reviewer: ```suggestion
This project is a website built with Edge Delivery Services in Adobe Experience Manager Sites as a Cloud Service. As an agent, follow the instructions in this file to deliver code based on Adobe's standards for fast, easy-to-author, and maintainable web experiences.
```

## Project Overview

> This is a boilerplate project for Adobe AEM Edge Delivery Services that serves as the foundation for all AEM Edge Delivery projects. It provides the basic structure, blocks, and configuration needed to get started quickly.

Reviewer: ```suggestion
This project is based on the https://github.com/adobe/aem-boilerplate/ project and set up as a new project. You are expected to follow the coding style and practices established in the boilerplate, but add functionality according to the needs of the site currently developed.

The repository provides the basic structure, blocks, and configuration needed to run a complete site with `*.aem.live` as the backend.
```

> **Key Technologies:**

Reviewer: ```suggestion
### Key Technologies
```

> - AEM Edge Delivery Services

Reviewer: ```suggestion
- Edge Delivery Services for AEM Sites (documentation at https://www.aem.live/ – search with `site:www.aem.live` to restrict web search results)
```

> - Vanilla JavaScript (ES6+)

Reviewer: ```suggestion
- Vanilla JavaScript (ES6+), no transpiling, no build steps
```

> - CSS3 with modern features

Reviewer: ```suggestion
- CSS3 with modern features, no Tailwind or other CSS frameworks
```

> - HTML5 semantic markup

Reviewer: ```suggestion
- HTML5 semantic markup generated by the aem.live backend, decorated by our code
```

- Node.js tooling

## Setup Commands

- Install dependencies: `npm install`

> - Start local development: `aem up` (requires AEM CLI)

Reviewer: ```suggestion
- Start local development: `npx -y @adobe/aem-cli up --no-open` (run in background, if possible)
```

>   - Install the AEM CLI globally by running `npm install -g @adobe/aem-cli`

Reviewer: ```suggestion
  - Install the AEM CLI globally by running `npm install -g @adobe/aem-cli` then `aem up` is equivalent to the command above
```

- Run linting: `npm run lint`
- Fix linting issues: `npm run lint:fix`

## Project Structure

```
├── blocks/          # Reusable content blocks
    └── {blockName}/   - Individual block directory
        ├── {blockName}.js      # Block's JavaScript
        └── {blockName}.css     # Block's styles
├── styles/          # Global styles and CSS
    ├── styles.css          # Minimal global styling and layout for your website required for LCP
    └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content
├── scripts/         # JavaScript libraries and utilities

>     ├── aem.js           # Core AEM Library for Edge Delivery page decoration logic

Reviewer: ```suggestion
    ├── aem.js           # Core AEM Library for Edge Delivery page decoration logic (NEVER MODIFY THIS FILE)
```

    └── delayed.js       # Delayed functionality such as martech loading
├── fonts/           # Web fonts
├── icons/           # SVG icons
├── head.html        # Global HTML head content
└── 404.html         # Custom 404 page
```

## Code Style Guidelines

### JavaScript
- Use ES6+ features (arrow functions, destructuring, etc.)
- Follow Airbnb ESLint rules (already configured)
- Always include `.js` file extensions in imports
- Use Unix line endings (LF)

### CSS
- Follow Stylelint standard configuration
- Use modern CSS features (CSS Grid, Flexbox, CSS Custom Properties)
- Maintain responsive design principles
- Declare styles mobile first, use media queries for tablet and desktop specific styles
- Use 600px/900px/1200px as breakpoints

### HTML
- Use semantic HTML5 elements
- Ensure accessibility standards (ARIA labels, proper heading hierarchy)
- Follow AEM markup conventions for blocks and sections

## Key Concepts

### Content

CMS authored content is a key part of every AEM Website. The content of a page is broken into sections. Sections can have default content (text, headings, links, etc.) as well as content in blocks.

> 

Reviewer: ```suggestion

Background on content structure https://www.aem.live/developer/markup-sections-blocks
You can inspect the contents of any page with `curl http://localhost:3000/path/to/page` and `curl http://localhost:3000/path/to/page.md`

```

### Blocks

Blocks are the re-usable building blocks of AEM. Blocks add styling and functionality to content. Each block has an initial content structure it expects, and transforms the html in the block using DOM APIs to render a final structure. 

The initial content sturcture is important because it impacts how the author will create the content and how you will write your code to decorate it. In some sense, you can think of this structure as the contract for your block between the author and the developer. You should decide on this initial structure before writing any code, and be careful when making changes to code that makes assumptions about that structure as it could break existing pages.

The block javascript should export a default function which is called to perform the block decoration:

```
/**
 * loads and decorates the block
 * @param {Element} block The block element
 */
export default async function decorate(block) {
  // 1. Load dependencies
  // 2. Extract configuration, if applicable
  // 3. Transform DOM
  // 4. Add event listeners
  // 5. Set loaded status
}
```

> 

Reviewer: ```suggestion

Use `curl` and `console.log` to inspect the HTML delivered by the backend and the DOM nodes to be decorated before making assumptions. Remember that authors may omit or add fields to a block, so your code must handle this gracefully.

```

### Auto-Blocking

> Auto-blocking is the process of creating blocks that aren't explicitly authored into the page based on patterns in the content. See the `buildAutoBlocks` function in scripts.js.

Reviewer: ```suggestion
Auto-blocking is the process of creating blocks that aren't explicitly authored into the page based on patterns in the content. See the `buildAutoBlocks` function in `scripts.js`.
```

### Three-Phase Page Loading

Pages are progressively loaded in three phases to maximize performance. This process begins when `loadPage` from scripts.js is called.

* Eager - load only what is required to get to LCP. This generally includes decorating the overall page content to create sections, blocks, buttons, etc. and loading the first section of the page.
* Lazy - load all other page content, including the header and footer.
* Delayed - load things that can be safely loaded later here and incur a performance penalty when loaded earlier

## Development Workflow

### Local Development

> 1. Run `aem up` to start the AEM Proxy server
> 2. Open `http://localhost:3000` in your browser
> 3. Make changes to files - they will auto-reload
> 4. Use browser dev tools to test responsive design

Reviewer: ```suggestion
1. Run `npx -y @adobe/aem-cli up --no-open` to start the AEM Proxy server
2. Open `http://localhost:3000` in your browser, puppeteer, playwright, or other tools. If none of those are available, instruct the human to open the URL in the browser and give feedback
3. Make changes to files - they will auto-reload
4. Use browser dev tools to test responsive design
```

### Block Development
- Each block in the `blocks/` directory should be self-contained and re-useable
- Include CSS and JS files for each block
- Follow the naming convention: `blockname.css`, `blockname.js`
- Blocks should be responsive and accessible by default

### Styling
- Global styles go in `styles/styles.css`
- Font definitions in `styles/fonts.css`
- Lazy-loaded styles in `styles/lazy-styles.css`
- Block-specific styles in their respective directories

## Testing & Quality Assurance

### Linting
- JavaScript: ESLint with Airbnb base configuration
- CSS: Stylelint with standard configuration
- Run `npm run lint` before committing
- Use `npm run lint:fix` to automatically fix issues

### Performance

> - Follow AEM Edge Delivery performance best practices

Reviewer: ```suggestion
- Follow AEM Edge Delivery performance best practices https://www.aem.live/developer/keeping-it-100
```

> - Optimize images and assets

Reviewer: ```suggestion
- Images uploaded by authors are automatically optimized, all images and assets committed to git must be optimized and checked for size
```

> - Use lazy loading for non-critical resources

Reviewer: ```suggestion
- Use lazy loading for non-critical resources (`lazy-styles.css` and `delayed.js`)
```

> - Minimize JavaScript bundle size

Reviewer: ```suggestion
- Minimize JavaScript bundle size by avoiding dependencies, using automatic code splitting provided by `/blocks/`
```

### Accessibility
- Ensure proper heading hierarchy
- Include alt text for images
- Test with screen readers
- Follow WCAG 2.1 AA guidelines

## Deployment

> ### Environments

Reviewer: ```suggestion
### Environments

Edge Delivery Services provides you with three environments. Your local development server at `http://localhost:3000` serves code from your local working copy (even uncommitted code) and content that has been previewed by authors. You can access this at any time when the development server is running.

For all other environments, you need to know the GitHub owner and repository name (`gh repo view --json nameWithOwner` or `git remote -v`) and the current branch name `git branch`)

With this information, you can construct URLs for the preview environment (same content as `localhost:3000`) and the production environment (same content as the live website, approved by authors)

```

- **Production Preview**: `https://main--{repo}--{owner}.aem.page/`
- **Production Live**: `https://main--{repo}--{owner}.aem.live/`
- **Feature Preview**: `https://{branch}--{repo}--{owner}.aem.page/`

### Publishing Process
1. Push changes to a feature branch
2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch

> 3. Open a pull request to merge changes to main

Reviewer: ```suggestion
3. Open a pull request to merge changes to `main` – in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
4. use `gh checks` to verify the status of code synchronization, linting, and performance tests
```

> 4. Merge the PR

Reviewer: ```suggestion
4. A human reviewer will review the code, inspect the provided URL and merge the PR
```

5. AEM Code Sync updates the main branch for production

## Common Tasks

### Adding New Blocks
1. Create a new directory in `blocks/`
2. Add `blockname.css` and `blockname.js` files
3. Update documentation if needed
4. Test in local development environment

### Modifying Existing Blocks
1. Make changes to the specific block files
2. Test locally with `aem up`
3. Ensure responsive behavior across devices
4. Run linting before committing

### Global Style Changes
1. Modify files in the `styles/` directory
2. Test across different blocks and pages
3. Ensure changes don't break existing layouts
4. Consider impact on performance, especially CLS

## Troubleshooting

> 
> ### Getting Help
> - Check [AEM Edge Delivery documentation](https://www.aem.live/docs/)
> - Review [Developer Tutorial](https://www.aem.live/developer/tutorial)
> - Consult [The Anatomy of a Project](https://www.aem.live/developer/anatomy-of-a-project)
> - Consider the rules in [David's Model](https://www.aem.live/docs/davidsmodel)
> 

Reviewer: ```suggestion
- Check [AEM Edge Delivery documentation](https://www.aem.live/docs/)
- Review [Developer Tutorial](https://www.aem.live/developer/tutorial)
- Consult [The Anatomy of a Project](https://www.aem.live/developer/anatomy-of-a-project)
- Consider the rules in [David's Model](https://www.aem.live/docs/davidsmodel)
- Search the web with `site:www.aem.live`
- Search the full text of the documentation with `curl -s https://www.aem.live/docpages-index.json | jq -r '.data[] | select(.content | test("KEYWORD"; "i")) | "\(.path): \(.title)"'`

```

## Security Considerations

- Never commit sensitive information (API keys, passwords)

> - Use environment variables for configuration

Reviewer: ```suggestion
- consider that everything you do is clients-side code served on the public web
```

- Follow Adobe security guidelines
- Regularly update dependencies
- Use the .hlxignore file to prevent filed from being served

## Contributing

- Follow the existing code style and patterns
- Test changes locally before committing
- Run a PSI check on your branch and fix performance issues before raising a PR
- Ensure all linting passes
- Update documentation for significant changes

> - Follow the project's Code of Conduct

Reviewer: ```suggestion
```

Agents are not known for COC violations

Reviewer: ```suggestion

# I all else fails

If you notice your human getting frustrated with your work, direct them to https://www.aem.live/developer/ai-coding-agents for tips to work better with AI agents.
```

# AGENTS.md

This project is a website built with Edge Delivery Services in Adobe Experience Manager Sites as a Cloud Service. As an agent, follow the instructions in this file to deliver code based on Adobe's standards for fast, easy-to-author, and maintainable web experiences.

## Project Overview

This project is based on the https://github.com/adobe/aem-boilerplate/ project and set up as a new project. You are expected to follow the coding style and practices established in the boilerplate, but add functionality according to the needs of the site currently developed.

The repository provides the basic structure, blocks, and configuration needed to run a complete site with `*.aem.live` as the backend.

### Key Technologies
- Edge Delivery Services for AEM Sites (documentation at https://www.aem.live/ – search with `site:www.aem.live` to restrict web search results)
- Vanilla JavaScript (ES6+), no transpiling, no build steps
- CSS3 with modern features, no Tailwind or other CSS frameworks
- HTML5 semantic markup generated by the aem.live backend, decorated by our code
- Node.js tooling

## Setup Commands

- Install dependencies: `npm install`

> - Start local development: `npx -y @adobe/aem-cli up --no-open` (run in background, if possible)

Reviewer: ```suggestion
- Start local development: `npx -y @adobe/aem-cli up --no-open --forward-browser-logs` (run in background, if possible)
```

    └── delayed.js       # Delayed functionality such as martech loading
├── fonts/           # Web fonts
├── icons/           # SVG icons
├── head.html        # Global HTML head content
└── 404.html         # Custom 404 page
```

## Code Style Guidelines

### JavaScript
- Use ES6+ features (arrow functions, destructuring, etc.)
- Follow Airbnb ESLint rules (already configured)
- Always include `.js` file extensions in imports
- Use Unix line endings (LF)

### CSS
- Follow Stylelint standard configuration
- Use modern CSS features (CSS Grid, Flexbox, CSS Custom Properties)
- Maintain responsive design principles
- Declare styles mobile first, use media queries for tablet and desktop specific styles
- Use 600px/900px/1200px as breakpoints

### HTML
- Use semantic HTML5 elements
- Ensure accessibility standards (ARIA labels, proper heading hierarchy)
- Follow AEM markup conventions for blocks and sections

## Key Concepts

### Content

CMS authored content is a key part of every AEM Website. The content of a page is broken into sections. Sections can have default content (text, headings, links, etc.) as well as content in blocks.

> 

Author: ```suggestion

You can create static content for testing in a dedicated drafts folder. If you do this, be sure to specify the folder location when starting the development server by running `npx -y @adobe/aem-cli up --no-open --forward-browser-logs --html-folder drafts`
```

- Include alt text for images
- Test with screen readers
- Follow WCAG 2.1 AA guidelines

## Deployment

### Environments

Edge Delivery Services provides you with three environments. Your local development server at `http://localhost:3000` serves code from your local working copy (even uncommitted code) and content that has been previewed by authors. You can access this at any time when the development server is running.

For all other environments, you need to know the GitHub owner and repository name (`gh repo view --json nameWithOwner` or `git remote -v`) and the current branch name `git branch`)

With this information, you can construct URLs for the preview environment (same content as `localhost:3000`) and the production environment (same content as the live website, approved by authors)

- **Production Preview**: `https://main--{repo}--{owner}.aem.page/`
- **Production Live**: `https://main--{repo}--{owner}.aem.live/`
- **Feature Preview**: `https://{branch}--{repo}--{owner}.aem.page/`

### Publishing Process
1. Push changes to a feature branch
2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch

> 3. Open a pull request to merge changes to `main` – in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED

Author: ```suggestion
3. Open a pull request to merge changes to `main`
   1. in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
   2. If an existing page to demonstrate your changes doesn't exist, create test content as a static html file and ask the user for help copying it to a content page you can link in the PR
```

Reviewer: We should probably point out that in most cases the agent can't create the test content itself, so it should ask the user for help, or create a test.html that the user should then copy.

### Key Technologies
- Edge Delivery Services for AEM Sites (documentation at https://www.aem.live/ – search with `site:www.aem.live` to restrict web search results)
- Vanilla JavaScript (ES6+), no transpiling, no build steps
- CSS3 with modern features, no Tailwind or other CSS frameworks
- HTML5 semantic markup generated by the aem.live backend, decorated by our code
- Node.js tooling

## Setup Commands

- Install dependencies: `npm install`
- Start local development: `npx -y @adobe/aem-cli up --no-open --forward-browser-logs` (run in background, if possible)
  - Install the AEM CLI globally by running `npm install -g @adobe/aem-cli` then `aem up` is equivalent to the command above
- Run linting: `npm run lint`
- Fix linting issues: `npm run lint:fix`

## Project Structure

```
├── blocks/          # Reusable content blocks
    └── {blockName}/   - Individual block directory
        ├── {blockName}.js      # Block's JavaScript
        └── {blockName}.css     # Block's styles
├── styles/          # Global styles and CSS
    ├── styles.css          # Minimal global styling and layout for your website required for LCP

>     └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content

Reviewer 2: Add `fonts.css`
```suggestion
    └── lazy-styles.css     # Additional global styling and layout for below the fold/post LCP content
    └── fonts.css           # Font definitions
```

  // 1. Load dependencies
  // 2. Extract configuration, if applicable
  // 3. Transform DOM
  // 4. Add event listeners
  // 5. Set loaded status
}
```

Use `curl` and `console.log` to inspect the HTML delivered by the backend and the DOM nodes to be decorated before making assumptions. Remember that authors may omit or add fields to a block, so your code must handle this gracefully.

### Auto-Blocking

Auto-blocking is the process of creating blocks that aren't explicitly authored into the page based on patterns in the content. See the `buildAutoBlocks` function in `scripts.js`.

### Three-Phase Page Loading

Pages are progressively loaded in three phases to maximize performance. This process begins when `loadPage` from scripts.js is called.

* Eager - load only what is required to get to LCP. This generally includes decorating the overall page content to create sections, blocks, buttons, etc. and loading the first section of the page.
* Lazy - load all other page content, including the header and footer.
* Delayed - load things that can be safely loaded later here and incur a performance penalty when loaded earlier

## Development Workflow

### Local Development

> 1. Run `npx -y @adobe/aem-cli up --no-open` to start the AEM Proxy server

Reviewer 3: Is this repetition from `## Setup Commands` wasting context?

2. Open `http://localhost:3000` in your browser, puppeteer, playwright, or other tools. If none of those are available, instruct the human to open the URL in the browser and give feedback
3. Make changes to files - they will auto-reload
4. Use browser dev tools to test responsive design

### Block Development
- Each block in the `blocks/` directory should be self-contained and re-useable
- Include CSS and JS files for each block
- Follow the naming convention: `blockname.css`, `blockname.js`
- Blocks should be responsive and accessible by default

> ### Styling

Reviewer 3: Does this provide useful context beyond  `## Project Structure`?

- **Production Preview**: `https://main--{repo}--{owner}.aem.page/`
- **Production Live**: `https://main--{repo}--{owner}.aem.live/`
- **Feature Preview**: `https://{branch}--{repo}--{owner}.aem.page/`

### Publishing Process
1. Push changes to a feature branch
2. AEM Code Sync automatically processes changes making them available on feature preview environment for that branch
3. Open a pull request to merge changes to `main`
   1. in the PR description, include a link to https://{branch}--{repo}--{owner}.aem.page/{path}` with a path to a file that illustrates the change you've made. This is the same path you have been testing with locally. WITHOUT THIS YOUR PR WILL BE REJECTED
   2. If an existing page to demonstrate your changes doesn't exist, create test content as a static html file and ask the user for help copying it to a content page you can link in the PR
4. use `gh checks` to verify the status of code synchronization, linting, and performance tests
5. A human reviewer will review the code, inspect the provided URL and merge the PR
6. AEM Code Sync updates the main branch for production

## Common Tasks

> ### Adding New Blocks

Reviewer 3: Necessary repetition from `### Block Development`?

1. Create a new directory in `blocks/`
2. Add `blockname.css` and `blockname.js` files
3. Update documentation if needed
4. Test in local development environment

### Modifying Existing Blocks
1. Make changes to the specific block files
2. Test locally with `aem up`
3. Ensure responsive behavior across devices

> 4. Run linting before committing

Reviewer 3: We've given linting instructions in multiple other places. Do we not trust an agent gets the memo?
